### PR TITLE
Create parser code for pilar logstash plugin

### PR DESCRIPTION
**Pull request by yash-dani**
This adds the parser functionality from PILAR into the plugin. Features include:

- checking for dynamic log entries based on trigram and bigram frequency analysis 
- generating log templates and events as parser output
- branch coverage for the code to check for dynamic tokens

### CODE REVIEW

File: lib/logstash/filters/parser.rb [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+require 'digest'
+
+# Define a Parser class for processing tokens and generating templates.
+class Parser
+  def initialize(tokens_list, single_dict, double_dict, tri_dict, threshold)
+    @tokens_list = tokens_list
+    @single_dict = single_dict
+    @double_dict = double_dict
+    @tri_dict = tri_dict
```
Comment (aaronabraham311) [inline lib/logstash/filters/parser.rb:11]: Why do we need to initialize this? We should grab it from `gram_dict` whenever we need. Let's pass in `gram_dict` as a parameter

File: lib/logstash/filters/parser.rb [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    f_str = f_score.to_s
+    if @entropy_dict.include?(f_str)
+      @entropy_dict[f_str] += 1
+    else
+      @entropy_dict[f_str] = 1
+    end
+  end
+
+  def is_dynamic(tokens, dynamic_index, index)
+    f = 0
```
Comment (aaronabraham311) [inline lib/logstash/filters/parser.rb:28]: Can we use a more descriptive variable name here

File: spec/filters/pilar_spec.rb [hunk 3]
```diff
@@ -1,7 +0,0 @@
-# frozen_string_literal: true
```
Comment (aaronabraham311) [inline spec/filters/pilar_spec.rb:1]: We can just leave this file, we will need to fill it in later

File: lib/logstash/filters/parser.rb [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+    if index.zero?
+      f = 1
+    elsif index == 1
+      singlegram = tokens[index - 1]
+      doublegram = "#{tokens[index - 1]}^#{tokens[index]}"
```
Comment (aaronabraham311) [inline lib/logstash/filters/parser.rb:34]: I have the same token concatenation logic going in in `preprocessor.rb`. Should we factor out all the token concat into some common function then? That way, we can standardize and make sure the logic remains the same.

File: lib/logstash/filters/parser.rb [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+      f = if @double_dict.include?(doublegram) && @single_dict.include?(singlegram)
+            @double_dict[doublegram].to_f / @single_dict[singlegram]
+          else
+            0
+          end
+
+      update_entropy_dict(f)
+    else
+      if dynamic_index.include?(index - 2)
```
Comment (aaronabraham311) [inline lib/logstash/filters/parser.rb:44]: I am really confused what `dyanmic_index` is, ping me when you have this commented

File: lib/logstash/filters/parser.rb [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        singlegram = tokens[index - 1]
+        doublegram = "#{tokens[index - 1]}^#{tokens[index]}"
+        f = if @double_dict.include?(doublegram) && @single_dict.include?(singlegram)
+              @double_dict[doublegram].to_f / @single_dict[singlegram]
+            else
+              0
+            end
```
Comment (aaronabraham311) [inline lib/logstash/filters/parser.rb:51]: Repeat work as lines 36-40? Can we factor this out?

File: lib/logstash/filters/parser.rb [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+      else
+        doublegram = "#{tokens[index - 2]}^#{tokens[index - 1]}"
+        trigram = "#{doublegram}^#{tokens[index]}"
+        f = if @tri_dict.include?(trigram) && @double_dict.include?(doublegram)
+              @tri_dict[trigram].to_f / @double_dict[doublegram]
+            else
+              0
+            end
```
Comment (aaronabraham311) [inline lib/logstash/filters/parser.rb:59]: You keep repeating the same logic, just with different index calculations. Can we refactor this?

File: lib/logstash/filters/parser.rb [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+      end
+      update_entropy_dict(f)
+    end
+    f <= @threshold
+  end
+
+  def gram_checker(tokens)
```
Comment (aaronabraham311) [inline lib/logstash/filters/parser.rb:66]: I understand that the original PILAR code used this name. Do you think there's a better name for this function? I find that the naming conventions in the original code is pretty unclear

File: lib/logstash/filters/parser.rb [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+  def update_entropy_dict(f_score)
+    f_str = f_score.to_s
+    if @entropy_dict.include?(f_str)
+      @entropy_dict[f_str] += 1
+    else
+      @entropy_dict[f_str] = 1
+    end
+  end
+
+  def is_dynamic(tokens, dynamic_index, index)
```
Comment (aaronabraham311) [inline lib/logstash/filters/parser.rb:27]: Which argument in this function are we checking for being "dynamic"? Can you rename the function to better indicate this?

File: lib/logstash/filters/parser.rb [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+            end
+      end
+      update_entropy_dict(f)
+    end
+    f <= @threshold
+  end
+
+  def gram_checker(tokens)
+    dynamic_index = []
+    if tokens.length >= 2
```
Comment (aaronabraham311) [inline lib/logstash/filters/parser.rb:68]: If `tokens` is an empty array or even one token, why do we want `dynamic_index` to remain empty?

File: lib/logstash/filters/parser.rb [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    template
+  end
+
+  # TODO: WE ARE NOW OUTPUTTING SOMETHING DIFFERENT AND THE INPUTS OF THE FUNCTION ARE DIFFERENT
+  # THIS NOW MAPS BETTER TO STREAMING FOR NOW
+  def parse
+    template_dict = {}
+
+    event_string = String.new("EventId,EventTemplate\n")
+    template_string = String.new("EventTemplate,Occurrences\n")
```
Comment (aaronabraham311) [inline lib/logstash/filters/parser.rb:97]: Why do we need this?

File: lib/logstash/filters/parser.rb [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+
+    @tokens_list.each do |tokens|
+      dynamic_index = gram_checker(tokens)
```
Comment (aaronabraham311) [inline lib/logstash/filters/parser.rb:100]: can we rename this variable, `dynamic_index` implies it's one index, but its actually returning multiple

File: lib/logstash/filters/parser.rb [hunk 13]
```diff
@@ -0,0 +1,118 @@
+# frozen_string_literal: true
+
+require 'digest'
+
+# Define a Parser class for processing tokens and generating templates.
+class Parser
+  def initialize(tokens_list, single_dict, double_dict, tri_dict, threshold)
+    @tokens_list = tokens_list
```
Comment (aaronabraham311) [inline lib/logstash/filters/parser.rb:8]: Why are we passing this in on initialization? Don't we want to pass this on calls to `parse`?

File: lib/logstash/filters/parser.rb [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+
+    event_string = String.new("EventId,EventTemplate\n")
+    template_string = String.new("EventTemplate,Occurrences\n")
+
+    @tokens_list.each do |tokens|
+      dynamic_index = gram_checker(tokens)
+      template = template_generator(tokens, dynamic_index)
+
+      # Remove specific characters from the template
+      template.gsub!(/[,'"]/, '') # TODO: SANITY CHECK IF THIS IS EQUIVALENT TO THE OLD REGEX, P SURE IT IS
```
Comment (aaronabraham311) [inline lib/logstash/filters/parser.rb:104]: Why?

File: lib/logstash/filters/parser.rb [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+
+      id = Digest::MD5.hexdigest(template)[0...4]
+
+      event_string << "e#{id},#{template}\n"
```
[the remark this answers is not in the document]
Comment (aaronabraham311) [reply lib/logstash/filters/parser.rb:108]: Can you explain why we need this if we are converting to stream-based?

File: lib/logstash/filters/parser.rb [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+
+      template_dict[template] = template_dict.fetch(template, 0) + 1
+    end
+
+    template_dict.each do |tmp, count|
+      template_string << "#{tmp},#{count}\n"
+    end
+    [event_string, template_string]
```
[the remark this answers is not in the document]
Comment (aaronabraham311) [reply lib/logstash/filters/parser.rb:116]: Ohh I see, you are not sure what to output yet. Nvm

**Review by aaronabraham311 — CHANGES_REQUESTED** [review]
Comments are sorely needed because I don't really follow the logic. Let me know when you have it commented it out.

Also would highly encourage you to think about refactoring the code if necessary. The original naming & structure in the Python code might not have been optimal. This seems to be a direct translation from Python, but we might want to change that up.

Also, take a look at my `GramDict` code. I think you should add some functions to get the `singledict`, `trigram_dict` etc. in the `GramDict` class rather than passing in the entire dictionary into `Parser` for two reasons.

1. The dictionaries are updated on the fly. If we are initializing `Parser` in the `intialize` function, we can't really pass in these three dictionaries

2. Seems a lot more clean to pass in the instance of `GramDict` than dissembling the instance to get the three inner dictionaries

Let me know if you want clarity on this.

File: lib/logstash/filters/parser.rb (hunk 15 above)
[the remark this answers is not in the document]
Comment (yash-dani) [reply lib/logstash/filters/parser.rb:108]: same reason as below

**Review by aaronabraham311 — APPROVED** [review]
Comments clarified a lot, thanks. Feel free to merge. If there's any other refactors/cleanup, just chuck it into [this ticket](https://www.notion.so/yashdani/Parser-cleanup-4789e90b56b04e80b26a11cf365e2e51?pvs=4)

**Pull request merged**